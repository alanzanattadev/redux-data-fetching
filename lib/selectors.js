"use babel";
// @flow

import { parse } from "graphql";
import { fromJS, Map, List } from "immutable";

type ImmutableSelector = Array<string | number | ({ type: string } & Object)>;

function getImmutableSelector(graphQLQuery: string): ImmutableSelector {
  let ast = parse(graphQLQuery);
  let query = ast.definitions[0];
  let rootField = query.selectionSet.selections[0];
  let rootFieldName = rootField.name.value;
  let rootFieldParamID;
  let immutableSelector: ImmutableSelector = [];
  immutableSelector.push(rootFieldName);
  let idArg;
  if (
    rootField.arguments.length > 0 &&
    (idArg = rootField.arguments.find(arg => arg.name.value == "id"))
  ) {
    rootFieldParamID = idArg.value.value;
    immutableSelector.push({ type: "id", value: rootFieldParamID });
  }
  return immutableSelector;
}

function convertIDsToIndex(
  data: Object,
  selector: ImmutableSelector,
): ImmutableSelector {
  return selector
    .reduce(
      (reduction: { rootData: any, newSelector: any }, value: any) => {
        if (
          reduction.newSelector.count() > 0 &&
          reduction.newSelector.last() === null
        )
          return reduction;
        else {
          if (typeof value == "string" || typeof value == "number") {
            return {
              rootData: reduction.rootData.get(value),
              newSelector: reduction.newSelector.push(value),
            };
          } else {
            if (value.type == "id") {
              let selectedIndex = reduction.rootData.findIndex(
                item => item.get("id") == value.value,
              );
              if (selectedIndex == -1) selectedIndex = null;
              return {
                rootData: selectedIndex
                  ? reduction.rootData.get(selectedIndex)
                  : null,
                newSelector: reduction.newSelector.push(selectedIndex),
              };
            } else {
              return reduction;
            }
          }
        }
      },
      { rootData: data, newSelector: List() },
    )
    .newSelector.toJS();
}

export function selectData(data: Object, query: string) {
  let selectorWithIDs = getImmutableSelector(query);
  let selector = convertIDsToIndex(data, selectorWithIDs);
  return Map().set(selectorWithIDs[0], data.getIn(selector, null));
}
