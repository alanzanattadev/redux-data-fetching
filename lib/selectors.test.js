import {
  printSchema,
  graphql,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString,
  GraphQLFloat,
  GraphQLList,
  GraphQLInt,
  GraphQLBoolean,
  GraphQLID,
  getNullableType,
} from "graphql";
import { convertsTypesSchemaToSelectorSchema } from "./selectors";
import { Record, Map, List } from "immutable";

describe("Selectors", () => {
  const UserType = new GraphQLObjectType({
    name: "User",
    fields: () => ({
      id: { type: GraphQLInt },
      name: { type: GraphQLString },
      friends: { type: new GraphQLList(UserType) },
    }),
  });
  const Power = new GraphQLObjectType({
    name: "Power",
    fields: {
      possible_entities: { type: UserType },
    },
  });
  const Skill = new GraphQLObjectType({
    name: "Skill",
    fields: () => ({
      id: { type: GraphQLInt },
      power: { type: GraphQLInt },
      target: { type: UserType },
    }),
  });
  const littleSchema = new GraphQLSchema({
    query: new GraphQLObjectType({
      name: "RootQueryType",
      fields: {
        users: {
          args: { gender: { type: GraphQLString } },
          type: new GraphQLList(UserType),
          resolve() {
            return [];
          },
        },
        skills: {
          type: new GraphQLList(Skill),
          resolve() {
            return [];
          },
        },
        user: {
          type: UserType,
          resolve() {
            return {};
          },
        },
      },
    }),
  });

  describe("convertsTypesSchemaToSelectorSchema", () => {
    it("converts simple schema to selector schema", () => {
      const subject = convertsTypesSchemaToSelectorSchema(littleSchema);

      expect(subject).toMatchSnapshot();
    });

    it("resolves to result of the needs", done => {
      const selectorSchema = convertsTypesSchemaToSelectorSchema(littleSchema);
      const queryHash = "hash";
      const User = Record({ id: null, name: null, friends: List() });
      const Progress = Record({ percent: 0, status: "WAITING" });
      const Result = Record({
        results: Map({ byQuery: Map(), byEntity: Map() }),
        progress: new Progress(),
      });
      const user1 = new User({ id: 1, name: "Alan", friends: List.of(2, 3) });
      const user2 = new User({
        id: 2,
        name: "Sophie",
        friends: List.of(1),
      });
      const user3 = new User({
        id: 3,
        name: "Antoine",
        friends: List.of(1),
      });
      const data = Map({
        queries: Map({
          [queryHash]: new Result({
            results: Map({
              byEntity: Map({ User: [2, 3] }),
              byQuery: Map({ users: [1, 3], females: [2] }),
            }),
          }),
        }),
        entities: Map({
          User: Map({ 1: user1, 2: user2, 3: user3 }),
        }),
      });
      const contextValue = {
        db: data,
        queryHash: queryHash,
      };
      graphql(
        selectorSchema,
        `{ users(gender: "male") { id, friends { id, friends { id, name }} }, females: users(gender: "female") { id, friends { id, name } } }`,
        null,
        contextValue,
      ).then(result => {
        expect(result.errors).toBeUndefined();
        expect(result.data).toMatchSnapshot();
        done();
      });
    });

    it("throws when db is not passed in contextValue", done => {
      const selectorSchema = convertsTypesSchemaToSelectorSchema(littleSchema);
      graphql(
        selectorSchema,
        `{ users(gender: "male") { id, friends { id } } }`,
      ).then(result => {
        expect(result.errors).toMatchSnapshot();
        done();
      });
    });
  });

  describe("selectData", () => {});
});
