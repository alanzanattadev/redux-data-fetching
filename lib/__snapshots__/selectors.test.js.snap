// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selectors convertsQueryToSelectorQuery converts name of fragment type in selector name 1`] = `
Object {
  "definitions": Array [
    Object {
      "directives": Array [],
      "kind": "OperationDefinition",
      "loc": Object {
        "end": 101,
        "start": 11,
      },
      "name": null,
      "operation": "query",
      "selectionSet": Object {
        "kind": "SelectionSet",
        "loc": Object {
          "end": 101,
          "start": 11,
        },
        "selections": Array [
          Object {
            "alias": null,
            "arguments": Array [],
            "directives": Array [],
            "kind": "Field",
            "loc": Object {
              "end": 89,
              "start": 25,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 30,
                "start": 25,
              },
              "value": "users",
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "loc": Object {
                "end": 89,
                "start": 31,
              },
              "selections": Array [
                Object {
                  "directives": Array [],
                  "kind": "FragmentSpread",
                  "loc": Object {
                    "end": 55,
                    "start": 47,
                  },
                  "name": Object {
                    "kind": "Name",
                    "loc": Object {
                      "end": 55,
                      "start": 50,
                    },
                    "value": "infos",
                  },
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "loc": Object {
                    "end": 75,
                    "start": 71,
                  },
                  "name": Object {
                    "kind": "Name",
                    "loc": Object {
                      "end": 75,
                      "start": 71,
                    },
                    "value": "name",
                  },
                  "selectionSet": null,
                },
              ],
            },
          },
        ],
      },
      "variableDefinitions": null,
    },
    Object {
      "directives": Array [],
      "kind": "FragmentDefinition",
      "loc": Object {
        "end": 164,
        "start": 113,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 127,
          "start": 122,
        },
        "value": "infos",
      },
      "selectionSet": Object {
        "kind": "SelectionSet",
        "loc": Object {
          "end": 164,
          "start": 136,
        },
        "selections": Array [
          Object {
            "alias": null,
            "arguments": Array [],
            "directives": Array [],
            "kind": "Field",
            "loc": Object {
              "end": 152,
              "start": 150,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 152,
                "start": 150,
              },
              "value": "id",
            },
            "selectionSet": null,
          },
        ],
      },
      "typeCondition": Object {
        "kind": "NamedType",
        "loc": Object {
          "end": 135,
          "start": 131,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 135,
            "start": 131,
          },
          "value": "UserSelector",
        },
      },
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 173,
    "start": 0,
  },
}
`;

exports[`Selectors convertsTypesSchemaToSelectorSchema converts simple schema to selector schema 1`] = `
GraphQLSchema {
  "_directives": Array [
    GraphQLDirective {
      "args": Array [
        Object {
          "astNode": undefined,
          "defaultValue": undefined,
          "description": "Included when true.",
          "name": "if",
          "type": "Boolean!",
        },
      ],
      "astNode": undefined,
      "description": "Directs the executor to include this field or fragment only when the \`if\` argument is true.",
      "locations": Array [
        "FIELD",
        "FRAGMENT_SPREAD",
        "INLINE_FRAGMENT",
      ],
      "name": "include",
    },
    GraphQLDirective {
      "args": Array [
        Object {
          "astNode": undefined,
          "defaultValue": undefined,
          "description": "Skipped when true.",
          "name": "if",
          "type": "Boolean!",
        },
      ],
      "astNode": undefined,
      "description": "Directs the executor to skip this field or fragment when the \`if\` argument is true.",
      "locations": Array [
        "FIELD",
        "FRAGMENT_SPREAD",
        "INLINE_FRAGMENT",
      ],
      "name": "skip",
    },
    GraphQLDirective {
      "args": Array [
        Object {
          "astNode": undefined,
          "defaultValue": "No longer supported",
          "description": "Explains why this element was deprecated, usually also including a suggestion for how to access supported similar data. Formatted in [Markdown](https://daringfireball.net/projects/markdown/).",
          "name": "reason",
          "type": "String",
        },
      ],
      "astNode": undefined,
      "description": "Marks an element of a GraphQL schema as no longer supported.",
      "locations": Array [
        "FIELD_DEFINITION",
        "ENUM_VALUE",
      ],
      "name": "deprecated",
    },
  ],
  "_implementations": Object {},
  "_mutationType": undefined,
  "_queryType": "SelectorsRootQueryType",
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "Int": "Int",
    "SelectorsRootQueryType": "SelectorsRootQueryType",
    "SkillSelector": "SkillSelector",
    "String": "String",
    "UserSelector": "UserSelector",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
  },
  "astNode": null,
}
`;

exports[`Selectors convertsTypesSchemaToSelectorSchema resolves to result of the needs 1`] = `
Object {
  "females": Array [
    Object {
      "friends": Array [
        Object {
          "id": 1,
          "name": "Alan",
        },
      ],
      "id": 2,
    },
  ],
  "users": Array [
    Object {
      "friends": Array [
        Object {
          "friends": Array [
            Object {
              "id": 1,
              "name": "Alan",
            },
          ],
          "id": 2,
        },
        Object {
          "friends": Array [
            Object {
              "id": 1,
              "name": "Alan",
            },
          ],
          "id": 3,
        },
      ],
      "id": 1,
    },
    Object {
      "friends": Array [
        Object {
          "friends": Array [
            Object {
              "id": 2,
              "name": "Sophie",
            },
            Object {
              "id": 3,
              "name": "Antoine",
            },
          ],
          "id": 1,
        },
      ],
      "id": 3,
    },
  ],
}
`;

exports[`Selectors convertsTypesSchemaToSelectorSchema resolves to result of the needs with fragments 1`] = `
Object {
  "females": Array [
    Object {
      "friends": Array [
        Object {
          "friends": Array [
            Object {
              "id": 2,
              "name": "Sophie",
            },
            Object {
              "id": 3,
              "name": "Antoine",
            },
          ],
          "id": 1,
          "name": "Alan",
        },
      ],
      "id": 2,
      "name": "Sophie",
    },
  ],
  "users": Array [
    Object {
      "friends": Array [
        Object {
          "friends": Array [
            Object {
              "id": 1,
              "name": "Alan",
            },
          ],
          "id": 2,
          "name": "Sophie",
        },
        Object {
          "friends": Array [
            Object {
              "id": 1,
              "name": "Alan",
            },
          ],
          "id": 3,
          "name": "Antoine",
        },
      ],
      "id": 1,
      "name": "Alan",
    },
    Object {
      "friends": Array [
        Object {
          "friends": Array [
            Object {
              "id": 2,
              "name": "Sophie",
            },
            Object {
              "id": 3,
              "name": "Antoine",
            },
          ],
          "id": 1,
          "name": "Alan",
        },
      ],
      "id": 3,
      "name": "Antoine",
    },
  ],
}
`;

exports[`Selectors convertsTypesSchemaToSelectorSchema throws when db is not passed in contextValue 1`] = `
Array [
  [GraphQLError: You have to pass database (entities and results) as db and graphql query as queryHash in contextValue],
]
`;
